File: mobile-engage/src/main/java/com/emarsys/mobileengage/iam/webview/IamWebViewClient.java
package com.emarsys.mobileengage.iam.webview;

import android.os.Handler;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.emarsys.core.util.Assert;

public class IamWebViewClient extends WebViewClient {

    private MessageLoadedListener listener;
    private Handler uiHandler;

    public IamWebViewClient(MessageLoadedListener listener, Handler uiHandler) {
        Assert.notNull(listener, "Listener must not be null!");
        Assert.notNull(uiHandler, "UiHandler must not be null!");
        this.listener = listener;
        this.uiHandler = uiHandler;
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        super.onPageFinished(view, url);
        uiHandler.post(new Runnable() {
            @Override
            public void run() {
                listener.onMessageLoaded();
            }
        });
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/iam/webview/IamWebViewClient.kt
package com.emarsys.mobileengage.iam.webview

import android.webkit.WebView
import android.webkit.WebViewClient
import com.emarsys.core.handler.ConcurrentHandlerHolder

class IamWebViewClient(
    private val listener: MessageLoadedListener,
    private val concurrentHandlerHolder: ConcurrentHandlerHolder
) : WebViewClient() {
    override fun onPageFinished(view: WebView, url: String) {
        super.onPageFinished(view, url)
        concurrentHandlerHolder.postOnMain { listener.onMessageLoaded() }
    }
}